lib/pg_party/adapter_decorator.rb

```diff
@@ -63,49 +109,138 @@ def detach_partition(parent_table_name, child_table_name)
       PgParty.cache.clear!
     end
 
-    private
+    def partitions_for_table_name(table_name, include_subpartitions:, _accumulator: [])
+      select_values(%[
+          SELECT pg_inherits.inhrelid::regclass::text
+          FROM pg_tables
+          INNER JOIN pg_inherits
+            ON pg_tables.tablename::regclass = pg_inherits.inhparent::regclass
+          WHERE pg_tables.tablename = #{quote(table_name)}
+                    ]).each_with_object(_accumulator) do |partition, acc|
+        acc << partition
+        next unless include_subpartitions
+
+        partitions_for_table_name(partition, include_subpartitions: true, _accumulator: acc)
+      end
+    end
 
-    def create_partition(table_name, type, partition_key, **options)
-      modified_options = options.except(:id, :primary_key, :template)
-      template         = options.fetch(:template, true)
-      id               = options.fetch(:id, :bigserial)
-      primary_key      = options.fetch(:primary_key) { calculate_primary_key(table_name) }
+    def parent_for_table_name(table_name, traverse: false)
+      parent = select_values(%[
+          SELECT pg_inherits.inhparent::regclass::text
+          FROM pg_tables
+          INNER JOIN pg_inherits
+            ON pg_tables.tablename::regclass = pg_inherits.inhrelid::regclass
+          WHERE pg_tables.tablename = #{quote(table_name)}
+      ]).first
+      return parent if parent.nil? || !traverse
+
+      while (parents_parent = parent_for_table_name(parent)) do
+        parent = parents_parent
+      end
 
-      validate_primary_key(primary_key)
+      parent
+    end
+
+    def add_index_on_all_partitions(table_name, column_name, in_threads: nil, **options)
```

Comment:
This is really only necessary for PG 10, correct?

Reply:
The #1 use case for this method is to provide an algorithm for creating properly referenced indices on partitioned tables using CONCURRENTLY (which postgres 11/12 does not allow to be used when you create an index on a partitioned (parent) tables. If you don't do this and you add an index to a partitioned table in Postgres 11+ it will lock all the partitions while it builds the index. So, with this method, you can have a table with 100s or even 1000s of partitions, including multiple levels of subpartitions, and you can add an index to all of them conveniently without any table locks and take advantage of postgres 11+'s index propagation so if you add a new partition later it will pick up the index automatically.

See my other comment further down for more info.

`in_threads:` takes an integer and with CONCURRENTLY will spread the index creation across multiple connections which is helpful if you have a whole lot of partitions, especially if some of them are rather large tables. Concurrent index creation is rather slow...

Reply:
Do you think it would be overkill to validate that the number of threads isn't higher than the connection pool? I could go either way on that. Curious what you think

Reply:
Yes, it seems like a good idea. I added a check in https://github.com/webandtech/pg_party/commit/194486cafc08dba50b9e119513eebfa03105435e which requires in_threads to be lower than connection pool size (there should at least one additional connection for the parent thread)